### PR TITLE
Schedule

### PR DESCRIPTION
**Pull request by NikczemnyOkularnik**
@Kasia9311 @Crazyayane07 @NikczemnyOkularnik

### CODE REVIEW

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule
🏅 Granted 12 XP for @Kasia9311

**Comment by ziobron** [comment]
⏱ 3 XP extra for @Kasia9311 for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @Kasia9311

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule
🏅 Granted 12 XP for @Crazyayane07

**Comment by ziobron** [comment]
⏱ 3 XP extra for @Crazyayane07 for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @Crazyayane07

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule
🏅 Granted 12 XP for @NikczemnyOkularnik

**Comment by ziobron** [comment]
⏱ 3 XP extra for @NikczemnyOkularnik for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @NikczemnyOkularnik

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule-bonus
🏅 Granted 10 XP for @Kasia9311

**Comment by ziobron** [comment]
⏱ 3 XP extra for @Kasia9311 for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @Kasia9311

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule-bonus
🏅 Granted 10 XP for @Crazyayane07

**Comment by ziobron** [comment]
⏱ 3 XP extra for @Crazyayane07 for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @Crazyayane07

**Comment by ziobron** [comment]
✅ CI Passed - Task: schedule-bonus
🏅 Granted 10 XP for @NikczemnyOkularnik

**Comment by ziobron** [comment]
⏱ 3 XP extra for @NikczemnyOkularnik for on-time delivery

**Comment by ziobron** [comment]
🏅 +1 XP for cooperation goes to @NikczemnyOkularnik

File: homework/schedule/schedule.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//         std::chrono::duration<double> time_elapsed = std::chrono::system_clock::now() - clock;
+//         return time_elapsed.count();
+//     };
+//     while(duration_time() < demanded_time){}
+//     Func(s, d);
+// }
+template <typename Func, typename... Args>
+void schedule(Func function, std::chrono::seconds seconds, Args... arguments)
+{   
+    auto demanded_time = seconds.count();
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:39]: const

File: homework/schedule/schedule.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    auto clock = std::chrono::system_clock::now();
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:40]: const

File: homework/schedule/schedule.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    auto duration_time = [clock](){
+        std::chrono::duration<double> time_elapsed = std::chrono::system_clock::now() - clock;
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:42]: put it on one line :)

Comment (nauka-programowania-MA) [reply homework/schedule/schedule.hpp:42]: return (now - clock).count()

File: homework/schedule/schedule.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        return time_elapsed.count();
+    };
+    while(duration_time() < demanded_time){}
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:45]: use std::this_thread::sleep_for or sleep_until instead loop.

File: homework/schedule/schedule.hpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    function(arguments ...);
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:46]: if you use variadict template, use also std::forward<Args>(arguments).... to perfect forward all args.

**Review by nauka-programowania-MA — APPROVED** [review]

**Review by nauka-programowania-MA — APPROVED** [review]

File: homework/schedule/schedule.hpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+//     };
+//     while(duration_time() < demanded_time){}
+//     Func(s, d);
+// }
+template <typename Func, typename... Args>
+void schedule(Func function, std::chrono::seconds seconds, Args... arguments)
+{   
+    std::this_thread::sleep_for(seconds);
+    function(std::forward<Args>(arguments) ...);
+}
```
Comment (nauka-programowania-MA) [inline homework/schedule/schedule.hpp:42]: Invoke args into the function using perfect forwarding :)

std::invoke(f, std::forward<Args>(args)...);

**Review by nauka-programowania-MA — APPROVED** [review]

**Pull request closed without merging**